lxd/storage/drivers/generic.go

```diff
@@ -225,28 +217,59 @@ func genericBackupUnpack(d Driver, vol Volume, snapshots []string, srcData io.Re
 	}
 
 	for _, snapName := range snapshots {
-		// Prepare tar arguments.
-		args := append(tarArgs, []string{
-			"-",
-			"--recursive-unlink",
-			"--xattrs-include=*",
-			"--strip-components=3",
-			"-C", vol.MountPath(), fmt.Sprintf("backup/snapshots/%s", snapName),
-		}...)
-
-		// Extract snapshots.
-		srcData.Seek(0, 0)
-		err = shared.RunCommandWithFds(srcData, nil, "tar", args...)
+		err = vol.MountTask(func(mountPath string, op *operations.Operation) error {
+			// Prepare tar arguments.
+			args := append(tarArgs, []string{
+				"-",
+				"--recursive-unlink",
+				"--xattrs-include=*",
+				"--strip-components=3",
+				"-C", mountPath, fmt.Sprintf("backup/snapshots/%s", snapName),
+			}...)
+
+			// Extract snapshot.
+			srcData.Seek(0, 0)
+			err = shared.RunCommandWithFds(srcData, nil, "tar", args...)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		}, op)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		snapVol, err := vol.NewSnapshot(snapName)
 		if err != nil {
 			return nil, nil, err
 		}
 
-		fullSnapshotName := GetSnapshotVolumeName(vol.name, snapName)
-		snapVol := NewVolume(d, d.Name(), vol.volType, vol.contentType, fullSnapshotName, vol.config, vol.poolConfig)
 		err = d.CreateVolumeSnapshot(snapVol, op)
 		if err != nil {
 			return nil, nil, err
 		}
+		revert.Add(func() { d.DeleteVolumeSnapshot(snapVol, op) })
+
+		// Remove files unpacked into the volume directory so the next volume can be unpacked into it.
+		err = wipeDirectory(vol.MountPath())
```

Comment:
@stgraber I wanted to draw your attention to this, and check whether the BTRFS driver also has the same issue.

When unpacking the backup tarball, the old approach was to create an empty volume, unpack each snapshot into it, create a snapshot of that volume, and then repeat for each snapshot and the subsequent primary volume.

However I reasoned that this would potentially be problematic as files that existed in an earlier snapshot but removed from later ones (or the main volume) would be left in the main volume (and interim snapshots) because the main unpack volume wasn't being wiped between each unpack.

Is this a valid concern, and do you think this is an acceptable approach to fix it?

Thanks

Reply:
I thought we passed a flag to tar to have it wipe everything clean.

Reply:
Yeah, we do, we pass `--recursive-unlink`

Reply:
@stgraber ah OK, TIL about that option, have removed. :)